### PR TITLE
Add developer docs for the codegen stage

### PR DESCRIPTION
**Pull request by EmilyBourne**
- Add documentation for the codegen stage.
- Make minor changes to the C and Fortran printers to conform to the documentation
    - Use `get_temporary_variable` where appropriate
    - Print `0` using `_print`
    - Use `DottedVariable` for neater printing in C
    - Move `stored_in_c_pointer` to the beginning of the class instead of in the middle of the `_print` functions
- Use `with` in the compilers (fixes bug where files were not closed)
- Remove unnecessary `xfail` marks
- Increase coverage
- Remove some references to `real` in the tests
- Activate tests for Python 3.11 in the master branch
- Remove dead code

### CODE REVIEW

File: developer_docs/codegen_stage.md
Comment (yguclu) [inline developer_docs/codegen_stage.md:257]: I would change "Python" with "NumPy" in this sentence.

File: developer_docs/codegen_stage.md
Comment (yguclu) [inline developer_docs/codegen_stage.md:267]: Capitalize: python -> Python

File: developer_docs/codegen_stage.md
Comment (yguclu) [inline developer_docs/codegen_stage.md:278]: Missing quotes for inline code: `Variables`

**Review by yguclu — CHANGES_REQUESTED** [review]
Excellent work @EmilyBourne! This is extremely useful!

File: pyccel/codegen/printing/ccode.py
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:326]: I agree with your name suggestion

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md:146]: I agree. Our `NumpyArraySize` returns an element of the tuple returned by NumPy's `shape` function. Do you have a suggestion for a better name? If so I think we should create an issue. It is a small change but it will touch lots of parts of the code so I would prefer to fix this in a separate PR

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md]: Fixed a88e2512

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md]: Fixed 28ba7ea1

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md]: Fixed 6230f785

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md]: Fixed 6230f785

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md]: Fixed 6230f785

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md]: Fixed 6230f785

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md]: Fixed 6230f785

File: pyccel/codegen/printing/ccode.py
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py]: Fixed abcd08d5 ?

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md:146]: `NumpyArraySize`->`NumpyArrayShapeElement`
`PyccelArraySize`->`NumpyArraySize`
?

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md]: Fixed a97e313c

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md]: Fixed a97e313c

File: pyccel/codegen/printing/ccode.py
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py]: And [e9c29eb](https://github.com/pyccel/pyccel/pull/1326/commits/e9c29ebcdf79a3919c647ca5cb9d543a72bed3e4)

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md:146]: Especially as every function where the name changes will want a new docstring

Comment (yguclu) [reply developer_docs/codegen_stage.md:146]: I agree with your name suggestion. And yes, we need a separate issue for this.

File: pyccel/codegen/printing/ccode.py
Comment (yguclu) [inline pyccel/codegen/printing/ccode.py:1040]: Does the returned string end with an empty space? According to the docstring that should not be necessary.

File: pyccel/codegen/printing/cwrappercode.py
Comment (yguclu) [inline pyccel/codegen/printing/cwrappercode.py:72]: Capitalize "C"?

**Review by yguclu — CHANGES_REQUESTED** [review]

File: pyccel/codegen/printing/ccode.py
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py]: Yes it does. Personally I think it is better for the context to decide whether the space is necessary but that is not how it is currently coded. I have changed this : 4e6aacf3

File: pyccel/codegen/printing/cwrappercode.py
Comment (EmilyBourne) [reply pyccel/codegen/printing/cwrappercode.py]: e0bc27ba

File: developer_docs/codegen_stage.md
Comment (EmilyBourne) [reply developer_docs/codegen_stage.md:146]: See #1329

**Comment by EmilyBourne** [comment]
@yguclu The remaining errors concern coverage. One cannot be helped and I have opened an issue for the other

**Comment by yguclu** [comment]
> @yguclu The remaining errors concern coverage. One cannot be helped and I have opened an issue for the other

How do you want to proceed, then?

**Comment by EmilyBourne** [comment]
> > @yguclu The remaining errors concern coverage. One cannot be helped and I have opened an issue for the other
> 
> How do you want to proceed, then?

It's up to you. I'm not sure it's worth holding up docs to increase coverage. I would rather add the coverage in a separate PR in case it turns out that the untested code is broken. That said, in any case I would prefer that we had more reviews before merging. I think the input of newer developers is much more important for documentation PRs as it shows whether it's actually understandable, or only understandable for people who already have an idea what we're talking about :sweat_smile:

**Review by yguclu — APPROVED** [review]

**Pull request merged**
